File: ng/macos/Sources/JiaoWoXuanCore/CourseFilter.swift
import Foundation

/// 课程目录的筛选条件。排序由表格列头决定(见 CourseRow 的排序键),这里只负责筛选。
public struct CourseFilter: Equatable, Sendable {
    public static let allCategories = "全部"

    public var query = ""
    public var category = CourseFilter.allCategories
    public var onlyOpen = false
    public var onlyUnassigned = false

    public init() {}

    public var isActive: Bool {
        category != Self.allCategories || onlyOpen || onlyUnassigned
    }

    public func apply(to courses: [CourseRow]) -> [CourseRow] {
        let needle = query.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        guard !needle.isEmpty || isActive else { return courses }
        return courses.filter { course in
            if category != Self.allCategories && course.category != category { return false }
            if onlyOpen && course.availability != .open { return false }
            if onlyUnassigned && course.group != nil { return false }
            if !needle.isEmpty && !course.searchText.contains(needle) { return false }
            return true
        }
    }
}

// MARK: - 列头排序用的键(都是非可选类型,可直接用于 KeyPathComparator)

extension CourseRow {
    public var sortKch: String { kch ?? "" }
    public var sortGroup: String { group ?? "" }

    /// 剩余名额;无法解析时排在最后(升序)/最前(降序)。
    public var sortRemainingSeats: Int {
        let parts = seatText.split(separator: "/").map { Int($0.trimmingCharacters(in: .whitespaces)) }
        guard parts.count == 2, let selected = parts[0], let capacity = parts[1] else { return Int.min }
        return capacity - selected
    }

    /// 已选 < 有空位 < 已满 < 未知。
    public var sortStatus: Int {
        if chosen { return 0 }
        switch availability {
        case .open: return 1
        case .full: return 2
        case .unknown: return 3
        }
    }

    /// 只有 rated 参与分数排序;未评分记为 -1。
    public var sortRating: Double { rating.sortScore ?? -1 }
}

File: ng/windows/JiaoWoXuan.Core/CourseFilter.cs
using System.Globalization;

namespace JiaoWoXuan.Core;

/// 可点击排序的课程表列。
public enum CourseColumn { Title, Teachers, Schedule, Kch, Seats, Status, Rating, Group }

/// 列头排序状态;Column 为 null 表示目录原始顺序。
public sealed record CourseSortOrder(CourseColumn? Column = null, bool Descending = false)
{
    /// 点击列头:新列从升序开始,同一列在升序 → 降序 → 取消排序之间循环。
    public CourseSortOrder Toggle(CourseColumn column) =>
        Column != column ? new(column) : !Descending ? this with { Descending = true } : new();
}

/// 课程目录的筛选条件;排序由列头决定。
public sealed record CourseFilter
{
    public const string AllCategories = "全部";

    public string Query { get; init; } = "";
    public string Category { get; init; } = AllCategories;
    public bool OnlyOpen { get; init; }
    public bool OnlyUnassigned { get; init; }

    public bool IsActive => Category != AllCategories || OnlyOpen || OnlyUnassigned;

    static readonly CompareInfo ChinaCompare = CultureInfo.GetCultureInfo("zh-CN").CompareInfo;

    public List<CourseRow> Apply(IEnumerable<CourseRow> courses, CourseSortOrder? sort = null)
    {
        var needle = Query.Trim();
        var matches = needle.Length == 0 && !IsActive
            ? courses
            : courses.Where(course =>
                (Category == AllCategories || course.Category == Category)
                && (!OnlyOpen || course.Availability == Availability.Open)
                && (!OnlyUnassigned || course.Group is null)
                && (needle.Length == 0 || course.SearchText.Contains(needle, StringComparison.OrdinalIgnoreCase)));
        return Sorted(matches, sort ?? new CourseSortOrder());
    }

    public static List<CourseRow> Sorted(IEnumerable<CourseRow> courses, CourseSortOrder sort)
    {
        if (sort.Column is not { } column) return [.. courses];
        var text = Comparer<string>.Create(CompareText);
        // OrderBy 是稳定排序,相同键保持目录顺序。
        IOrderedEnumerable<CourseRow> ordered = column switch
        {
            CourseColumn.Title => Order(courses, c => c.Title, text, sort.Descending),
            CourseColumn.Teachers => Order(courses, c => c.Teachers, text, sort.Descending),
            CourseColumn.Schedule => Order(courses, c => c.FirstSchedule, text, sort.Descending),
            CourseColumn.Kch => Order(courses, c => c.Kch ?? "", text, sort.Descending),
            CourseColumn.Group => Order(courses, c => c.Group ?? "", text, sort.Descending),
            CourseColumn.Seats => Order(courses, c => c.SortRemainingSeats, Comparer<int>.Default, sort.Descending),
            CourseColumn.Status => Order(courses, c => c.SortStatus, Comparer<int>.Default, sort.Descending),
            _ => Order(courses, c => c.Rating.SortScore ?? -1, Comparer<double>.Default, sort.Descending),
        };
        return [.. ordered];
    }

    static IOrderedEnumerable<CourseRow> Order<TKey>(IEnumerable<CourseRow> rows, Func<CourseRow, TKey> key, IComparer<TKey> comparer, bool descending) =>
        descending ? rows.OrderByDescending(key, comparer) : rows.OrderBy(key, comparer);

    static int CompareText(string? left, string? right) =>
        ChinaCompare.Compare(left, right, CompareOptions.IgnoreCase | CompareOptions.IgnoreWidth | CompareOptions.NumericOrdering);
}
